### PR TITLE
Allow low-level use of Server

### PR DESCRIPTION
**Pull request by JoseKilo**
Close: #265 

Basic idea is being able to do something like this:

```
config = Config(app=make_app())
server = Server(config)
loop = asyncio.get_event_loop()
loop.run_until_complete(server.run())
```

This will allow to run other tasks in the same asyncio loop.

### CODE REVIEW

File: uvicorn/loops/uvloop.py [hunk 1]
```diff
@@ -4,6 +4,5 @@
 
 
 def uvloop_setup():
-    asyncio.get_event_loop().close()
```
Comment (JoseKilo) [inline uvicorn/loops/uvloop.py]: Not sure if this is actually needed. If I keep the line, it complains with `RuntimeError: Cannot close a running event loop` (using uvloop only).

File: uvicorn/main.py [hunk 2]
```diff
@@ -256,26 +256,37 @@ def __init__(self, config):
...
-        self.loop = config.loop_instance
+        if config.loop_instance is None:
+            config.setup_event_loop()
+
         self.logger = config.logger_instance
         self.lifespan = config.lifespan_class(config)
 
-        self.install_signal_handlers()
+        if config.install_signal_handlers:
+            self.install_signal_handlers()
```
Comment (JoseKilo) [inline uvicorn/main.py]: This check was missing, but the config property was there

**Comment by lovelydinosaur** [comment]
Okay, so I think we might want to get from A to B step by step here, as I think there's some related changes to make along the way.

First step: Let's *just* deal with splitting out the `Config.setup_event_loop`

I think we probably want to drop `loop_instance` on the config, and instead use `asyncio.get_event_loop()` everywhere instead. The only place where we can't do that is on the protocol classes, because for the sake of testing we want to be able to inject mock loop implementations there.

So, I think step one here is probably:

* Split out `Config.setup_event_loop()`
* Drop the `loop_instance` on the config completely.
* Use `asyncio.get_event_loop()` everywhere.
* In the protocol classes add an argument `_loop = None` on init, and use `self.loop = _loop or asyncio.get_event_loop()` - the test cases can then explicitly set the loop on the protocol.

Once we've got that in, then I think we should tackle the rest of the `main()` vs. `async run()` refactoring.

Does that seem reasonable?

**Comment by JoseKilo** [comment]
Yes, it makes sense. Let's start with the first one: #301

File: uvicorn/config.py [hunk 3]
```diff
@@ -83,7 +83,6 @@ def __init__(
         timeout_keep_alive=5,
         timeout_notify=30,
         callback_notify=None,
-        install_signal_handlers=True,
```
Comment (JoseKilo) [inline uvicorn/config.py:86]: Not sure if we want to remove it from the arguments (wouldn't be fully backwards compatible)

**Comment by JoseKilo** [comment]
I have rebased on top of master and left the `install_signal_handlers` change out.

File: uvicorn/config.py (hunk 3 above)
Comment (lovelydinosaur) [reply uvicorn/config.py:86]: It's not documented API, so we're good.
(We were just using it internally)

File: uvicorn/main.py [hunk 4]
```diff
@@ -401,12 +407,15 @@ def install_signal_handlers(self):
...
 
-    def handle_exit(self, sig, frame):
+    def stop(self):
         if self.should_exit:
             self.force_exit = True
         else:
             self.should_exit = True
 
+    def handle_exit(self, sig, frame):
+        self.stop()
```
Comment (lovelydinosaur) [inline uvicorn/main.py]: Let's not add this extra indirection.

File: uvicorn/main.py [hunk 5]
```diff
@@ -256,28 +256,34 @@ def __init__(self, config):
...
-        self.loop.run_until_complete(self.shutdown())
-        self.loop.stop()
+        await self.startup()
+        await self.main_loop()
+        await self.shutdown()
         self.logger.info("Finished server process [{}]".format(process_id))
 
+    @property
+    def loop(self):
+        return asyncio.get_event_loop()
```
Comment (lovelydinosaur) [inline uvicorn/main.py]: I think we should probably drop this, and just use `loop = asyncio.get_event_loop()` inside any methods that need it.

File: uvicorn/main.py [hunk 6]
```diff
@@ -256,28 +256,34 @@ def __init__(self, config):
         self.should_exit = False
         self.force_exit = False
 
-    def run(self):
+    def main(self):
+        self.config.setup_event_loop()
+
```
Comment (lovelydinosaur) [inline uvicorn/main.py:264]: This newline probably isn't meant to be here.

**Review by lovelydinosaur — CHANGES_REQUESTED** [review]
Looking really good!

I've put some minor comments in place to address. Other thing is: are we happy we have the naming right here?

Right now we have:

* `def uvicorn.run()`
* `def Server.main()`
* `async def Server.run()`

Which is a bit asymmetrical. Perhaps we should swap `run` and `main` here??? Or something else?

**Comment by lovelydinosaur** [comment]
Or perhaps:

`uvicorn.run()`
`server.run()`
`await server.serve()`

**Comment by lovelydinosaur** [comment]
Not sure exactly, but I guess it probably makes sense for the blocking server method name to match the blocking function name.

**Comment by JoseKilo** [comment]
Going with `uvicorn.run -> Server.run -> Server.serve`.

That way the previous Server API doesn't change, and the tests don't need to change.

**Review by lovelydinosaur — APPROVED** [review]

**Comment by lovelydinosaur** [comment]
Lovely, lovely stuff here! 🎉

I'm going to tweak the `__init__` imports to make `Server` and `Config` available from the main module. And also roll a new release.

At which point we will have the following:

Standard entry point:

```python
import uvicorn

uvicorn.run(app, **kwargs)
```

Run, blocking the current thread.

```python
import uvicorn

config = uvicorn.Config(app, **kwargs)
server = uvicorn.Server(config)
server.run()
```

Run using the event loop explictly.

```python
import asyncio
import uvicorn

config = uvicorn.Config(app, **kwargs)
server = uvicorn.Server(config)
loop = asyncio.get_event_loop()
loop.run_until_complete(server.serve())
```

In due course we should start documenting up the programatic interface.

**Comment by JoseKilo** [comment]
That's great ! Thank you for your help :)

**Pull request merged**